### ISSUE TITLE
SSH Proxy

### ISSUE THREAD

**Issue by yunge** [chobits/ngx_http_proxy_connect_module#36]
I want to proxy SSH connections on my server, port 443 to port 22. The reason is that sometimes the port 22 is not available.

The nginx conf is like this:

```
server{
    listen 443;
    server_name  ssh.mydomain;

    proxy_connect;
    proxy_connect_allow  all;
    proxy_connect_connect_timeout  15s;
    proxy_connect_read_timeout     30s;
    proxy_connect_send_timeout     30s;
    proxy_connect_address 127.0.0.1:22;

    # location / {
    #     proxy_pass http://127.0.0.1:22;
    # }

}
```

But I always get "The proxy server rejected connection request: Connection aborted." in Xshell. Any idea?

**Comment by chobits** [comment]
hi, could you provide nginx [debug log](http://nginx.org/en/docs/debugging_log.html) for issue.

**Comment by yunge** [comment]
Thanks for reply.

The error log:
2018/03/06 23:00:06 [debug] 16955#0: timer delta: 1
2018/03/06 23:00:06 [debug] 16955#0: worker cycle
2018/03/06 23:00:06 [debug] 16955#0: epoll timer: 29942
2018/03/06 23:00:06 [debug] 16955#0: epoll: fd:14 ev:0005 d:0000000002176A59
2018/03/06 23:00:06 [debug] 16955#0: timer delta: 87
2018/03/06 23:00:06 [debug] 16955#0: worker cycle
2018/03/06 23:00:06 [debug] 16955#0: epoll timer: 29855
2018/03/06 23:00:36 [debug] 16955#0: timer delta: 29876
2018/03/06 23:00:36 [debug] 16955#0: worker cycle
2018/03/06 23:00:36 [debug] 16955#0: epoll timer: -1

And the log of xshell:
Connecting to 192.168.31.195:9443...
Connection established.
To escape to local shell, press 'Ctrl+Alt+]'.
Connection closing...Socket close.

Connection closed by foreign host.

And btw, I found when I use the 443 port, other sites can't use the port, though they have different server names, how can I handle this problem?

**Comment by chobits** [comment]
Hi,

> I want to proxy SSH connections on my server, port 443 to port 22. 

I see. This module cannot meet your requirement. You should use a tcp proxy server (pure tcp stream proxy) to proxy SSH connection from port 443 to port 22. [ngx_stream_module](http://nginx.org/en/docs/stream/ngx_stream_proxy_module.html) can work.

This module is for HTTP-CONNECT tunnel. For more details, referer to https://en.wikipedia.org/wiki/HTTP_tunnel#HTTP_CONNECT_tunneling.

**Comment by neoadventist** [comment]
@yunge i'm trying to solve the same problem, can you ping when you've solved it!

**Comment by yunge** [comment]
Hi,
I used the http proxy mode of putty/xshell, so that they can send SNI information to nginx(SSH doesn't send domain name), and the putty does send CONNECT request to server. But anyway, thanks for your advice.

@neoadventist,
If you really need this function, may be you will be interested in "sslh", 
https://github.com/yrutschle/sslh
http://www.rutschle.net/tech/sslh/README.html
https://wiki.meurisse.org/wiki/sslh

**Comment by chobits** [comment]
hi @yunge

I have not used putty/xshell before, but I try to reproduce this issue. Could you provide your putty/xshell config (maybe need screenshot) . Then I can try to reproduce it.

Thanks

**Comment by adrian-007** [comment]
I have similar problem. I can run this module just fine on CentOS with nginx 1.12 rebuild from source RPM. When I configure server to run ssl server on port other than 443, it works fine, but once I run it like other sites, I get 400 Bad Request errors. I've tested it with a command:
``curl -x https://myserver.com https://github.com

**Comment by yunge** [comment]
@chobits 

Sorry that I missed your reply. My putty/xshell config are both default configs, you can download putty and I believe it's easy to reproduce it, thanks.

**Comment by jahoh** [comment]
I've tested it with the command:curl https://github.com/ -v -x 127.0.0.1:3128, but I got "curl(56) Proxy CONNECT aborted".

**Comment by anitakrueger** [comment]
@adrian-007 Have you found a way to solve this? I am getting the exact same 400 error only when I'm running a named virtual host on port 443 rather than another specific port.

**Comment by chobits** [comment]
@anitakrueger you can file a new issue, I guess your case is not the same one for this issue.

It helps me resolve the issue easily to know your nginx.conf and the request / data stream sent by your client,

**Comment by chobits** [comment]
For chrome-https-proxy-setting, you can referer this guide: https://github.com/chobits/ngx_http_proxy_connect_module/issues/22#issuecomment-346941271. ( This guide is not for `curl https://... -x ...`, this curl command is similar to chrome-http-proxy-setting, not to chrome-https-proxy-setting)

**Comment by adrian-007** [comment]
@chobits I decided to ditch this module and do it using vanilla nginx, that is I've reconfigured TLS servers to first dissect traffic, get SNI from request and depending on it's value and dispatch traffic to either squid (stream proxy) or to virtual server (running on local host now). I needed to do some trickery so that PHP sites properly get host name, but I got it working not only for squid, but also other services (i.e Matrix server).

18 cze 2019 17:06 Anita Krueger <notifications@github.com> napisał(a):

@adrian-007<https://github.com/adrian-007> Have you found a way to solve this? I am getting the exact same 400 error only when I'm running a named virtual host on port 443 rather than another specific port.
